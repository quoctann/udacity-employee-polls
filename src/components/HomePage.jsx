import { useState } from "react";
import { Button, ButtonGroup, Container } from "react-bootstrap";
import { connect } from "react-redux";
import QuestionList from "./QuestionList";

const HomePage = (props) => {
	const [newQuestion, doneQuestion] = ["New Question", "Done"];

	const [isNewQuestionView, setIsNewQuestionView] = useState(true);

	const handleToggleView = (value) => {
		setIsNewQuestionView(value);
	};

	const {
		questions,
		authedUser: { id },
	} = props;

	const [answered, unanswered] = [[], []];

	Object.values(questions)
		.sort((first, second) => second.timestamp - first.timestamp)
		.forEach((q) => {
			if ([...q.optionOne.votes, ...q.optionTwo.votes].includes(id)) {
				answered.push(q);
				return;
			}
			unanswered.push(q);
		});

	return (
		<Container>
			<ButtonGroup size="lg" className="my-2 mx-auto">
				<Button
					variant={isNewQuestionView ? "primary" : "secondary"}
					onClick={() => handleToggleView(true)}
				>
					{newQuestion}
				</Button>
				<Button
					variant={!isNewQuestionView ? "primary" : "secondary"}
					onClick={() => handleToggleView(false)}
				>
					{doneQuestion}
				</Button>
			</ButtonGroup>

			{isNewQuestionView ? (
				<QuestionList
					title={newQuestion}
					questionList={unanswered}
					isAnswered={false}
				/>
			) : (
				<QuestionList
					title={doneQuestion}
					questionList={answered}
					isAnswered={true}
				/>
			)}
		</Container>
	);
};

const mapStateToProps = ({ questions, authedUser }) => ({
	questions,
	authedUser,
});

export default connect(mapStateToProps)(HomePage);
