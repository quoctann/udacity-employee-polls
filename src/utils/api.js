import {
	_getUsers,
	_getQuestions,
	_saveQuestion,
	_saveQuestionAnswer,
	users,
} from "./_DATA.js";

export function getInitialData() {
	return Promise.all([_getUsers(), _getQuestions()]).then(
		([users, questions]) => ({ users, questions })
	);
}

export function saveQuestion(question) {
	return _saveQuestion(question);
}

export function saveQuestionAndAnswer({ authedUser, qid, answer }) {
	return _saveQuestionAnswer({ authedUser, qid, answer });
}

export function login(username, password) {
	return new Promise((resolve, reject) => {
		setTimeout(() => {
			const user = users[username];
			if (!user || user?.password !== password) {
				reject("Username or password incorrect!");
			}

			resolve(user);
		}, 500);
	});
}
