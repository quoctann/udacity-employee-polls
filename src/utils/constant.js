const APP_PATH = {
	ROOT: "/",
	LEADERBOARD: "/leaderboard",
	ADD_QUESTION: "/add",
	QUESTION_DETAIL: "/questions/:question_id",
	ERROR_PAGE: "/error",
};

Object.freeze(APP_PATH);

export { APP_PATH };
