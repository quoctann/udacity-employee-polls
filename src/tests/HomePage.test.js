import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import HomePage from "../components/HomePage";
import { questions } from "../utils/_DATA";

const mockStore = configureMockStore();

describe("Test HomePage component", () => {
	let store;

	beforeEach(() => {
		const initialState = {
			authedUser: {
				id: "tylermcginnis",
				password: "abc321",
				name: "Tyler McGinnis",
			},
			questions: questions,
		};
		store = mockStore(initialState);
	});

	it("Should render correct QuestionCard component", () => {
		const component = render(
			<Provider store={store}>
				<MemoryRouter>
					<HomePage />
				</MemoryRouter>
			</Provider>
		);

		expect(component).toMatchSnapshot();
	});
});
