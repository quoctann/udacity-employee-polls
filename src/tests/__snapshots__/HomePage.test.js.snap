// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HomePage component Should render correct QuestionCard component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="my-4 p-1 card"
        >
          <h3
            class="text-center card-header"
          >
            New Question
          </h3>
          <div
            class="card-body"
          >
            <div
              class="d-flex flex-wrap"
            >
              <div
                class="m-2 card"
                style="width: 18rem;"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title h5"
                  >
                    sarahedo
                  </div>
                  <p
                    class="card-text"
                  >
                    04 Mar 2017, 5:22:47 am
                  </p>
                  <div
                    class="d-grid"
                  >
                    <button
                      class="btn btn-outline-success"
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="m-2 card"
                style="width: 18rem;"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title h5"
                  >
                    tylermcginnis
                  </div>
                  <p
                    class="card-text"
                  >
                    24 Dec 2016, 6:42:47 pm
                  </p>
                  <div
                    class="d-grid"
                  >
                    <button
                      class="btn btn-outline-success"
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="m-2 card"
                style="width: 18rem;"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title h5"
                  >
                    mtsamis
                  </div>
                  <p
                    class="card-text"
                  >
                    14 Jul 2016, 2:02:47 pm
                  </p>
                  <div
                    class="d-grid"
                  >
                    <button
                      class="btn btn-outline-success"
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="m-2 card"
                style="width: 18rem;"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title h5"
                  >
                    sarahedo
                  </div>
                  <p
                    class="card-text"
                  >
                    29 Jun 2016, 9:21:12 am
                  </p>
                  <div
                    class="d-grid"
                  >
                    <button
                      class="btn btn-outline-success"
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="my-4 p-1 card"
        >
          <h3
            class="text-center card-header"
          >
            Done
          </h3>
          <div
            class="card-body"
          >
            <div
              class="d-flex flex-wrap"
            >
              <div
                class="m-2 card"
                style="width: 18rem;"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title h5"
                  >
                    mtsamis
                  </div>
                  <p
                    class="card-text"
                  >
                    01 May 2017, 2:16:07 am
                  </p>
                  <div
                    class="d-grid"
                  >
                    <button
                      class="btn btn-outline-success"
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="m-2 card"
                style="width: 18rem;"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title h5"
                  >
                    tylermcginnis
                  </div>
                  <p
                    class="card-text"
                  >
                    15 Mar 2017, 7:09:27 pm
                  </p>
                  <div
                    class="d-grid"
                  >
                    <button
                      class="btn btn-outline-success"
                      type="button"
                    >
                      Show
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="my-4 p-1 card"
      >
        <h3
          class="text-center card-header"
        >
          New Question
        </h3>
        <div
          class="card-body"
        >
          <div
            class="d-flex flex-wrap"
          >
            <div
              class="m-2 card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  sarahedo
                </div>
                <p
                  class="card-text"
                >
                  04 Mar 2017, 5:22:47 am
                </p>
                <div
                  class="d-grid"
                >
                  <button
                    class="btn btn-outline-success"
                    type="button"
                  >
                    Show
                  </button>
                </div>
              </div>
            </div>
            <div
              class="m-2 card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  tylermcginnis
                </div>
                <p
                  class="card-text"
                >
                  24 Dec 2016, 6:42:47 pm
                </p>
                <div
                  class="d-grid"
                >
                  <button
                    class="btn btn-outline-success"
                    type="button"
                  >
                    Show
                  </button>
                </div>
              </div>
            </div>
            <div
              class="m-2 card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  mtsamis
                </div>
                <p
                  class="card-text"
                >
                  14 Jul 2016, 2:02:47 pm
                </p>
                <div
                  class="d-grid"
                >
                  <button
                    class="btn btn-outline-success"
                    type="button"
                  >
                    Show
                  </button>
                </div>
              </div>
            </div>
            <div
              class="m-2 card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  sarahedo
                </div>
                <p
                  class="card-text"
                >
                  29 Jun 2016, 9:21:12 am
                </p>
                <div
                  class="d-grid"
                >
                  <button
                    class="btn btn-outline-success"
                    type="button"
                  >
                    Show
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="my-4 p-1 card"
      >
        <h3
          class="text-center card-header"
        >
          Done
        </h3>
        <div
          class="card-body"
        >
          <div
            class="d-flex flex-wrap"
          >
            <div
              class="m-2 card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  mtsamis
                </div>
                <p
                  class="card-text"
                >
                  01 May 2017, 2:16:07 am
                </p>
                <div
                  class="d-grid"
                >
                  <button
                    class="btn btn-outline-success"
                    type="button"
                  >
                    Show
                  </button>
                </div>
              </div>
            </div>
            <div
              class="m-2 card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  tylermcginnis
                </div>
                <p
                  class="card-text"
                >
                  15 Mar 2017, 7:09:27 pm
                </p>
                <div
                  class="d-grid"
                >
                  <button
                    class="btn btn-outline-success"
                    type="button"
                  >
                    Show
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
